### DISCUSSION TITLE
Support for FriendlyElec CM3588 or workaround for mutiple NVMe Drives

### DISCUSSION THREAD

**Discussion by DragonFlye** [Joshua-Riek/ubuntu-rockchip#859] [Ideas]
I'm using the NanoPC T6 24.04 LTS release for my new **[CM3588](https://wiki.friendlyelec.com/wiki/index.php/CM3588)**. 

Everything is working fine. However, the CM3588 has four NVMe m.2 slots (of which two occupied) and only the first is recognized.
Does the CM3588 need its own separate image or can this be solved another way?

Also the 'normal' USB2/3 ports don't recognize a common mouse/keyboard dongle. The USB-C port does.

**Comment by Joshua-Riek** [comment]
The CM3588 would need its own specific image.

**Comment by DragonFlye** [comment]
Would be _greatly_ appreciated if one could be available

Seems it's gotten quite popular since review on [Linus Tech Tips](https://www.youtube.com/watch?v=QsM6b5yix0U)

**Reply by Joshua-Riek** [reply]
Since I work on so many boards I can not financially afford them all. I will try and reach out to FriendlyElec, hopefully they will be able to send me a board for development.

**Reply by DragonFlye** [reply]
I understand. Not having the actual devices to test on.

I'm in a little group facing the same issues, but for development with the Android side of things. Trying to make something useful from the horrendous stock firmware. Getting full (and flawless) HW video decoding to work (including HEVC 10 bit) was the main priority. In the end Kodi can now play anything you throw at it. I'd like to achieve the same on Linux

Anything I can provide? Software wise?

**Comment by jani-hmg** [comment]
I also am wanting to run this distro on a cm3588
I did find their device tree in their repo, so it should be just a matter of adding that, right?
https://github.com/friendlyarm/kernel-rockchip/blob/nanopi6-v6.1.y/arch/arm64/boot/dts/rockchip/rk3588-nanopi6-rev09.dts

https://github.com/Joshua-Riek/ubuntu-rockchip/wiki/Ubuntu-24.04-LTS#changing-the-device-tree

**Reply by DragonFlye** [reply]
Forgive my ignorance, but I did use the NanoPC T6 image. If FriendlyElec claims the NanoPC T6 and CM3588 use the same device tree, it should already work. Or the rev09 changes haven't been implemented yet....

**Reply by jani-hmg** [reply]
The NanoPC T6 image does not contain the device tree I mentioned (it is from FriendlyElec repo).
I see their device tree lists 4 nvme devices, which does not seem to be in place in the NanoPC T6 device tree.
Also, the cm3588 compatible string is present in their device tree, suggesting that device tree is more accurate.

**Reply by Joshua-Riek** [reply]
A device tree needs to be added to the kernel for the CM3588 and U-Boot support needs to be added.

**Reply by DragonFlye** [reply]
Well jani, if you have any luck succeeding, please share your steps

**Reply by jani-hmg** [reply]
Was just about to get cracking, when I discovered there already is a dts for this in the repo:
https://github.com/Joshua-Riek/linux-rockchip/blob/noble/arch/arm64/boot/dts/rockchip/rk3588-nanopc-cm3588-nas.dts

Someone added it 3 months ago!

**Reply by jani-hmg** [reply]
I can verify that switching to that device-tree works fine. I have 2 nvmes installed and both show up just fine.
Have not done any more extensive testing so far, but atleast display and keyboard work too.
What I did was to install the nanopc t6 image, boot it up (shows only 1 nvme), then change the device tree according to the wiki to:

```
ubuntu@ubuntu:~$ sudo vi /etc/default/u-boot

U_BOOT_FDT="device-tree/rockchip/rk3588-nanopc-cm3588-nas.dtb"

ubuntu@ubuntu:~$ sudo u-boot-update
ubuntu@ubuntu:~$ reboot
```

Perhaps this could be added as a prebuilt image so we would not need to do the additional step of changing device tree.

**Reply by DragonFlye** [reply]
Can confirm it works!

Still some hurdles to overcome. Don't know what your intended use-case is, but I'd like to _try_ and make it a multipurpose SBC, running Kodi as main media player (acting also as launcher for browser and other apps) and Home Assistant (with other docker add-ons) in the background. For this to work I'll still need:

- Widevine support in Chromium
- A **stable** Kodi release, capable of HW decoding everything you'd throw at it

So far I've got it running, log-in straight into Kodi GUI and everything else running as services in the background. All that consuming very little power and resources

**Comment by Deiiji** [comment]
Hello, I use that nas board as well and I would love to see JR Ubuntu working on it (I have beta "pro" board with SPI ram on board and second normal one)

**Reply by DragonFlye** [reply]
Initially, the latest version 24.04 runs fine (with the modification). But it all depends on what you want to do with it once you've got it running. Enough for plain for desktop use and/or docker management.

Video playing/streaming isn't on par at this moment, making the CM3588 not as suited as a multi-purpose Media player/Server device.....for now.

**Reply by Deiiji** [reply]
I not need graphical output Jellyfin work just fine (you need recompile ffmpeg-jellyfin as it has a nasty bug in libs when you enable  hardware decoding). I need all 4 drives working and if possible, load oss directly from NVMe and not at all from emmc as this emmc will die fast with my kind of 
usage...

**Reply by Deiiji** [reply]
Your method don't work for me, I still have only first nvme detected, on both boards

**Reply by DragonFlye** [reply]
It should. My two nvme drives are recognized. However use 24.04. I didn't work on 22.04

**Reply by Deiiji** [reply]
I use 24.04 as well and each board has 4 nvme drives installed, one run dotnet api, other website and nas +jellyfin + icecast with liquidsoap. booting go well but still only first drive is seen. In fact is a second NVMe visible 3 others are not here, one usb 3 is not working as well, on both boards.

**Comment by tymercer** [comment]
Sorry for the dumb question here, but I can't figure out how to get things put onto the SPI-NOR device because it doesn't show up in /dev as it should.
I have gone through all of the posts, discussions, issues and such without any luck.

I downloaded all of the different device trees (3 i believe) and copied them to their location according to the post, but it still isn't there.

For the device tree edit needed in /etc/default/u-boot I'm not sure what that is referring to
U_BOOT_FDT="device-tree/rockchip/rk3588s-orangepi-5.dtb"

Where is the file that it's looking for, is that one of the files that was copied over from the zips?

Thanks for all of the work and effort, I'm just trying to figure out how to get this thing up and running and hopefully get a build of Opnsense running on here (but freebsd is being a turd).

**Reply by Deiiji** [reply]
Hello, regular cm3588 has no spi, only few had, if you want spi flash memory, you need add it by yourself to standard carier board, same for opi5 pro, by the way.
You cann wrire uboot on sd card or into mmc if your board has one then os on a usb or nvme disk.

**Reply by tymercer** [reply]
It has SPI-NOR onboard

**Comment by Deiiji** [comment]
Well on mine it show as mtdblock0 (beta pro), second one, regular one has no spi mounted

**Reply by tymercer** [reply]
yes, I'm trying to find out how to get mine to do so, with what file to use and where to put it, etc.

**Reply by Deiiji** [reply]
if you see spi then you follow install, it worked on my board , you need dd ubot

**Comment by Cicatr1x** [comment]
Im running the first Version of this Board with the FriendlyElec Debian Bookworm Image in eMMC. Could someone explain to my which Image i should use and how to get it onto the eMMC?

**Reply by Deiiji** [reply]
Hello, image to run is the one for NanoPC T6 but you will need edit config files to use  cm3588 device tree
All is in this post. If it don't work, liek for me, you will need rebuild device tree.

**Reply by Cicatr1x** [reply]
So probably better going for the armbian variant as there it should work better? I must admit im a total noob when it comes to building own kernels or playing around with device trees. Espacially im not sure how to install that on the internal eMMC.

https://github.com/armbian/linux-rockchip/pull/151

**Comment by mologie** [comment]
Just for future reference for anyone who stumbles across this, there is also a fan header onboard that won't function by default. The fan is properly configured, but disables by default.

An overlay is required to enable it:

```
mkdir -p /boot/overlay
cat >/boot/overlay/fan.dts <<EOF
/dts-v1/;
/plugin/;

/ {
    fragment@0 {
        target = <&fan>;
        __overlay__ {
            status = "okay";
        };
    };
};
EOF
dtc -I dts -O dtb -@ /boot/overlay/fan.dts -o /boot/overlay/fan.dtbo
```

Now edit `/etc/default/u-boot` to set (assuming this is your only overlay, there are none by default):
```
U_BOOT_FDT_OVERLAYS="fan.dtbo"
U_BOOT_FDT_OVERLAYS_DIR="/boot/overlay"
```

Finally, run `u-boot-update`. Reboot and the fan should spin and auto-adjust depending on CPU temp :).

**Reply by naohezi** [reply]
Any idea on how to manually set up speeds depending on temp?

**Reply by mologie** [reply]
@0xChups instead of just setting `status` in the override, also set your own `cooling-levels` and `rockchip,temp-trips`, check the upstream file for reference: https://github.com/Joshua-Riek/linux-rockchip/blob/noble/arch/arm64/boot/dts/rockchip/rk3588-nanopc-cm3588-nas.dts#L78

Not sure what the exact logic there is, but this looks like `cooling-levels` sets the PWM speed 0-255, and `rockchip,temp-trips` maps minimum temps to indices of that array (with an additional implicit zero index for temps below the first entry).

Good luck!
